3329 write kep for retriable and non-retriable pod failures for jobs (#1)
mimowo/enhancements: keps/sig-apps/3329-non-retriable-exit-codes/README.md

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

## Motivation

Running a large computational workload, comprising thousands of pods on
thousands of nodes requires usage of pod restart policies in order
to account for infrastructure failures.

Currently, kubernetes Job API offers a way to account for infrastructure
failures by setting `.backoffLimit > 0`. However, this mechanism intructs
job controller to restart all failed pods - regardless of the root cause
of the failures. However, in some scenarios this leads to unnecessary
restarts of many pods, resulting in a waste of time and computational
resources. What makes the restarts more expensive is the fact that the
failures may be encountered late in the execution time of a program.

It can often be determined by manual inspection of container exit codes

> that the root cause of a failure is in code of the executable and the

Reviewer: ```suggestion
Sometimes it can be determined from containers exit codes
that the root cause of a failure is in the executable and the
```

job is destined to fail regardless of the number of retries. However, since
the large workloads are often scheduled to run over night or over the

> weekend, there is no human assistance to terminate such a job early.

Reviewer: Add a paragraph for infrastructure failures, which are not always easy to determine based on exit codes.

And the motivation is that some platforms allow users to run arbitrary jobs and they don't know which exit codes to filter.

The need for solving the problem has been emphasized by the kubernetes
community in the issues, see: [#17244](https://github.com/kubernetes/kubernetes/issues/17244) and [#31147](https://github.com/kubernetes/kubernetes/issues/31147).

The lack of native k8s support for terminating jobs based on container exit codes

> motivated development of various framework-coupled retry policies:

Reviewer: ```suggestion
Some third-party frameworks have implemented retry policies for their pods:
```

- [TensorFlow Training (TFJob)](https://www.kubeflow.org/docs/components/training/tftraining/)
- [Argo workflow](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml) ([example](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml))

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals

Extension of Job API with user-friendly syntax to terminate jobs based on

> container exit codes.

Reviewer: or termination reasons

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

> Implementation of other indicators of non-retirable jobs such as: logs,

Reviewer: ```suggestion
- Implementation of other indicators of non-retriable jobs such as: logs,
```

stdout/stderr or duration of execution.

> 

Reviewer: remove this empty line

> Allowing for all pods of a failed indexed-job to finish execution

Reviewer: ```suggestion
- Allowing for all indexes of an indexed-job to execute when only one or a few indexes fail
```

[#109712](https://github.com/kubernetes/kubernetes/issues/109712).

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

> Extension of the Job API with a new flag which allows to configure a set of

Reviewer: ```suggestion
Extension of the Job API with a new field which allows to configure a set of
```

exit codes for containers which will be interpreted as non-retriable errors,
leading to the job termination.

Use the main job controller main loop to detect the fact that there is
a container which exited with a non-retriable exit code. In such case, the
job controller will set the job status as `Failed` with a dedicated condition
reflecting the fact that one of the containers failed with a non-retriable exit
codes. This will prevernt restart of new pods even if `.backoffLimit > 0`.

<!--
TODO: check what happens currently when a job is marked as terminated
-->

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

> ### User Stories (Optional)

Reviewer: Once the API is better defined, it is useful to include a snippet on how the API looks like for each story.

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

As a machine learning researcher, I run jobs comprising thousands
of long-running pods on a cluster comprising thousands of nodes. The jobs often
run at night or over weekend without any human monitoring. In order to account
for random infrastructure failures we define `.backoffLimit: 6` for the job.
However, a signifficant portion of the failures happen due to bugs in code.
Moreover, the failures may happen late during the program execution time. In
such case, restarting such a pod results in wasting a lot of computational time.

We would like to be able to automatically detect and terminate the jobs which are
failing due to the bugs in code of the executable, so that the computation resources
can be saved.

#### Story 2

As a service provider that offers computational resources to researchers I would like to

> have a mechanism which terminates jobs for which pods are failing due to user errors.

Reviewer: but allows infinite retries for infrastructure errors (pod terminations by kubelet, kube-scheduler, abnormal node termination, etc)

It is important that the mechanism is can automatically detect such jobs as I do not
have expertise in what the computations do. This is important in order to limit the

> costs of running the service.

Reviewer: ```suggestion
I do not have knowledge or influence over the executable that researchers run, so I don't know beforehand
which exit codes they might return. This is important in order to limit the
costs of running the service.
```

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!-- Increased complexity of the Job management, by interaction with some of
already exiting Job configuration options. To mitigate this, we introduce
minimal API which will allow to satisfy the main usage scenario.-->

> Container exit codes could be lost if the failed pod is garbage collected.

Reviewer: ```suggestion
The Pod status (which includes exit codes) could be lost if the failed pod is garbage collected.
```

First, it should be rather rare so the overall consequence of unnecessary pod restarts
will be limited. Second, we can prevent this by using the feature of
[job tracking with finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/).

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### JobSpec API

> To be discussed. Here are some variants:

Reviewer: Let's already pick one. I prefer variant 3.

Move the rest to "Alternatives"

#### Variant 1:

Single job-level field with range values. In this

```yaml
apiVersion: v1
kind: Job
spec:
  template:
    spec:
      containers:
      - name: main-job-container
        image: job-image
        command: ["./program"]
  backoffLimit: 3
  terminateOnExitCodes: 1-12,100-127
```

Pros:
- simple API extension satisfying the main usecase

Cons:
- the API may not be very flexible to extend in the future
- single range of exit codes for all containers

#### Variant 2

```yaml
apiVersion: v1
kind: Job
spec:
  template:
    spec:
      containers:
      - name: main-job-container
        image: job-image
        command: ["./program"]
      - name: monitoring-job-container
        image: job-monitoring
        command: ["./monitoring"]
  backoffLimit: 3
  backoffPolicy:
    terminateOnConditions:
    - exitCodes: 1-12,100-127
      containerName: main-job-container
```

> Here, we could make the `containerName` field optional. If not specified,

Reviewer: Include the type with all the fields to be added, like here https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2307-job-tracking-without-lingering-pods#api-changes
